File: backend/app/schemas/gpu.py
"""GPU resource mapping schemas."""

from pydantic import BaseModel, Field


class GPUDeviceMap(BaseModel):
    """A single node-level mapping entry for a GPU resource mapping."""

    node: str
    path: str
    id: str = ""
    subsystem_id: str | None = None
    iommu_group: int | None = None
    description: str | None = None
    is_mdev: bool = False


class GPUMappingPublic(BaseModel):
    """Public representation of a PVE PCI resource mapping (GPU)."""

    id: str = Field(description="Mapping logical ID (name)")
    description: str = ""
    maps: list[GPUDeviceMap] = Field(default_factory=list)


class GPUMappingDetail(GPUMappingPublic):
    """Detail view with usage information."""

    physical_gpu_count: int = Field(default=0, description="Estimated physical GPU count (by unique PCI bus)")
    device_count: int = Field(default=0, description="Total assignable device/VF slots")
    used_count: int = 0
    available_count: int = 0
    is_sriov: bool = Field(default=False, description="True if SR-IOV detected (multiple devices on same PCI bus)")
    has_mdev: bool = Field(default=False, description="True if any device uses mediated devices")
    total_vram_mb: int = Field(default=0, description="Total physical VRAM in MB")
    used_vram_mb: int = Field(default=0, description="Allocated VRAM in MB (sum of assigned vGPU/passthrough)")
    used_by: list["GPUUsageInfo"] = Field(default_factory=list)


class GPUUsageInfo(BaseModel):
    """Information about a VM using a GPU mapping."""

    vmid: int
    vm_name: str = ""
    node: str = ""
    status: str = ""
    mdev_type: str = ""
    allocated_vram_mb: int = 0


class GPUMappingsPublic(BaseModel):
    """List of GPU mappings."""

    data: list[GPUMappingDetail]
    count: int


class GPUMappingCreate(BaseModel):
    """Create a new PCI resource mapping."""

    id: str = Field(min_length=1, max_length=128, description="Mapping name")
    description: str = ""
    map: list[str] = Field(
        min_length=1,
        description="List of map entries, e.g. 'node=pve1,path=0000:01:00.0'",
    )


class GPUMappingUpdate(BaseModel):
    """Update an existing PCI resource mapping."""

    description: str | None = None
    map: list[str] | None = None


class GPUSummary(BaseModel):
    """A simplified GPU option for the application form selector."""

    mapping_id: str
    description: str = ""
    model: str = ""
    vram: str = ""
    node: str = ""
    physical_gpu_count: int = 1
    device_count: int = 1
    used_count: int = 0
    available_count: int = 1
    is_sriov: bool = False
    has_mdev: bool = False
    total_vram_mb: int = 0
    used_vram_mb: int = 0

File: frontend/src/services/gpu.ts
import type { CancelablePromise } from "@/client"
import { OpenAPI } from "@/client"
import { request as __request } from "@/client/core/request"

export type GPUDeviceMap = {
  node: string
  path: string
  id: string
  subsystem_id?: string | null
  iommu_group?: number | null
  description?: string | null
}

export type GPUUsageInfo = {
  vmid: number
  vm_name: string
  node: string
  status: string
  mdev_type: string
  allocated_vram_mb: number
}

export type GPUMappingDetail = {
  id: string
  description: string
  maps: GPUDeviceMap[]
  physical_gpu_count: number
  device_count: number
  used_count: number
  available_count: number
  is_sriov: boolean
  has_mdev: boolean
  total_vram_mb: number
  used_vram_mb: number
  used_by: GPUUsageInfo[]
}

export type GPUMappingsPublic = {
  data: GPUMappingDetail[]
  count: number
}

export type GPUSummary = {
  mapping_id: string
  description: string
  model: string
  vram: string
  node: string
  physical_gpu_count: number
  device_count: number
  used_count: number
  available_count: number
  is_sriov: boolean
  has_mdev: boolean
  total_vram_mb: number
  used_vram_mb: number
}

export type GPUMappingCreate = {
  id: string
  description?: string
  map: string[]
}

export const GpuService = {
  listMappings(): CancelablePromise<GPUMappingsPublic> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/gpu/mappings",
      errors: { 422: "Validation Error" },
    })
  },

  getMapping(mappingId: string): CancelablePromise<GPUMappingDetail> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/gpu/mappings/{mapping_id}",
      path: { mapping_id: mappingId },
      errors: { 422: "Validation Error" },
    })
  },

  createMapping(body: GPUMappingCreate): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "POST",
      url: "/api/v1/gpu/mappings",
      body,
      mediaType: "application/json",
      errors: { 422: "Validation Error" },
    })
  },

  deleteMapping(mappingId: string): CancelablePromise<{ message: string }> {
    return __request(OpenAPI, {
      method: "DELETE",
      url: "/api/v1/gpu/mappings/{mapping_id}",
      path: { mapping_id: mappingId },
      errors: { 422: "Validation Error" },
    })
  },

  listOptions(params?: {
    startAt?: string
    endAt?: string
  }): CancelablePromise<GPUSummary[]> {
    return __request(OpenAPI, {
      method: "GET",
      url: "/api/v1/gpu/options",
      query: {
        start_at: params?.startAt,
        end_at: params?.endAt,
      },
      errors: { 422: "Validation Error" },
    })
  },
}
